pkg/operations/kubernetesupgrade/upgrader.go

```diff
@@ -557,6 +596,94 @@ func (ku *Upgrader) generateUpgradeTemplate(upgradeContainerService *api.Contain
 	return templateMap, parametersMap, nil
 }
 
+func (ku *Upgrader) getLastVMNameInVMSS(ctx context.Context, resourceGroup string, vmScaleSetName string) (string, error) {
+	lastVMName := ""
+	for vmScaleSetVMsPage, err := ku.Client.ListVirtualMachineScaleSetVMs(ctx, resourceGroup, vmScaleSetName); vmScaleSetVMsPage.NotDone(); err = vmScaleSetVMsPage.Next() {
+		if err != nil {
+			return "", err
+		}
+
+		vms := vmScaleSetVMsPage.Values()
+		if vms != nil {
+			vm := vms[len(vms)-1]
+			lastVMName = *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName
+		}
+	}
+
+	if lastVMName == "" {
+		return "", fmt.Errorf("failed to get the last VM name in Scale Set %s", vmScaleSetName)
+	}
+
+	return lastVMName, nil
+}
+
+func (ku *Upgrader) copyCustomPropertiesToNewNode(client armhelpers.KubernetesClient, oldNodeName string, newNodeName string) error {
+	oldNode, err := client.GetNode(oldNodeName)
+	if err != nil {
+		return err
+	}
+
+	newNode, err := client.GetNode(newNodeName)
+	if err != nil {
+		return err
+	}
+
+	// copy additional custom annotations from old node to new node
+	if oldNode.Annotations != nil {
+		if newNode.Annotations == nil {
+			newNode.Annotations = map[string]string{}
+		}
+
+		for k, v := range oldNode.Annotations {
+			if _, ok := newNode.Annotations[k]; !ok {
+				newNode.Annotations[k] = strings.Replace(v, oldNodeName, newNodeName, -1)
+			}
+		}
+	}
+
+	// copy additional custom labels from old node to new node
+	if oldNode.Labels != nil {
+		if newNode.Labels == nil {
+			newNode.Labels = map[string]string{}
+		}
+
+		for k, v := range oldNode.Labels {
+			if _, ok := newNode.Labels[k]; !ok {
+				newNode.Labels[k] = strings.Replace(v, oldNodeName, newNodeName, -1)
+			}
+		}
+	}
+
+	// copy Taints from old node to new node
+	if oldNode.Spec.Taints != nil {
+		newNode.Spec.Taints = append([]v1.Taint{}, oldNode.Spec.Taints...)
+		for i := range newNode.Spec.Taints {
+			newNode.Spec.Taints[i].Value = strings.Replace(newNode.Spec.Taints[i].Value, oldNodeName, newNodeName, -1)
+		}
+	}
+
+	_, err = client.UpdateNode(newNode)
+
+	return err
+}
+
+func (ku *Upgrader) getKubernetesClient() (armhelpers.KubernetesClient, error) {
+	var kubeAPIServerURL string
+	getClientTimeout := 10 * time.Second
+
+	if ku.DataModel.Properties.HostedMasterProfile != nil {
```

Comment:
There is ku.DataModel.Properties.IsHostedMasterProfile()
Maybe we could add Properties.GetMasterFQDN() to wrap all this?

Reply:
+1